lib/jekyll/readers/layout_reader.rb

```diff
@@ -7,23 +7,30 @@ def initialize(site)
     end
 
     def read
-      layout_entries.each do |f|
-        @layouts[layout_name(f)] = Layout.new(site, layout_directory, f)
+      layout_entries.each do |directory, entries|
+        entries.each do |name|
+          @layouts[layout_name(name)] = Layout.new(site, directory, name)
+        end
       end
 
       @layouts
     end
 
-    def layout_directory
-      @layout_directory ||= (layout_directory_in_cwd || layout_directory_inside_source)
+    def layout_directories
+      @layout_directories ||= Array(site.config['layouts_dir']).map do |directory|
+        layout_directory_in_cwd(directory) || layout_directory_inside_source(directory)
+      end
     end
 
     private
 
     def layout_entries
-      entries = []
-      within(layout_directory) do
-        entries = EntryFilter.new(site).filter(Dir['**/*.*'])
+      entries = {}
```

Comment:
I suppose it is a bit less efficient, but using `reduce` or `inject` would be a bit more elegant, e.g.

```ruby
def layout_entries
  layout_directories.inject({}) do |layouts, directory|
    within(directory) do
      layouts[directory] = \
        EntryFilter.new(site).filter(Utils.safe_glob("*", ["**", "*.*"]))
    end 
    layouts
  end
end
```